contrib/lib/Xpex/extensions/lib/PEX/Xpex/doc/man/XpexCompose.man
.\" $Header: XpexCompose.man,v 2.5 91/09/11 16:04:33 sinyaw Exp $
.\"
.\"
.\" Copyright 1991 by Sony Microsystems Company, San Jose, California
.\" 
.\"                   All Rights Reserved
.\"
.\" Permission to use, modify, and distribute this software and its
.\" documentation for any purpose and without fee is hereby granted,
.\" provided that the above copyright notice appear in all copies and
.\" that both that copyright notice and this permission notice appear
.\" in supporting documentation, and that the name of Sony not be used
.\" in advertising or publicity pertaining to distribution of the
.\" software without specific, written prior permission.
.\"
.\" SONY DISCLAIMS ANY AND ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
.\" INCLUDING ALL EXPRESS WARRANTIES AND ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS, FOR A PARTICULAR PURPOSE. IN NO EVENT
.\" SHALL SONY BE LIABLE FOR ANY DAMAGES OF ANY KIND, INCLUDING BUT NOT
.\" LIMITED TO SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES RESULTING FROM
.\" LOSS OF USE, DATA OR LOSS OF ANY PAST, PRESENT, OR PROSPECTIVE PROFITS,
.\" WHETHER IN AN ACTION OF CONTRACT, NEGLIENCE OR OTHER TORTIOUS ACTION, 
.\" ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS 
.\" SOFTWARE.
.\"
.\" 
.TH XpexCompose 3PEX "$Revision: 2.5 $" "Sony Microsystems"
.AT
.SH "Name"
XpexComposeMatrix \- generate the composition of two homogeneous
2-D matrices.
.sp
XpexComposeMatrix2D \- generate the composition of two homogeneous
3-D matrices.
.sp
XpexComposeTransformationMatrix \- compose a 3D transformation
matrix, which is the composition of a specified matrix and a
transformation matrix defined by a fixed point, shift vector,
rotation angle, and scale factors.
.sp
XpexComposeTransformationMatrix2D \- compose a 2D transformation
matrix, which is the composition of a specified matrix and a 
transformation matrix defined by a fixed point, shift vector, 
rotation angle, and scale factors.
.sp
XpexComputeGeometricNormal \- compute geometric normal of the fill 
area set
.SH "Synopsis"
.nf
void XpexComposeMatrix(matrix_a, matrix_b, composed_matrix)
.br
	XpexMatrix matrix_a;  
.br
	XpexMatrix matrix_b;  
.br
	XpexMatrix composed_matrix;  /* RETURN */
.sp
void XpexComposeMatrix2D(matrix_a, matrix_b, composed_matrix)
.br
	XpexMatrix3X3 matrix_a;
.br
	XpexMatrix3X3 matrix_b;
.br
	XpexMatrix3X3 composed_matrix; /* RETURN */
.sp
void XpexComposeTransformationMatrix(transformation_matrix, 
.br
	fixed_point, shift_vector, rotation_angle_x, 
.br
	rotation_angle_y, rotation_angle_z, scale_vector, 
.br
	composed_matrix)
.br
	XpexMatrix transformation_matrix;
.br
	XpexCoord3D *fixed_point;  
.br
	XpexVector3D *shift_vector;
.br
	XpexFloat rotation_angle_x; 
.br
	XpexFloat rotation_angle_y;
.br
	XpexFloat rotation_angle_z;
.br
	XpexVector3D *scale_vector;
.br
	XpexMatrix composed_matrix; /* RETURN */
.sp
void XpexComposeTransformationMatrix2D(transformation_matrix, 
.br
	fixed_point, shift_vector, rotation_angle, scale_vector, 
.br
	composed_matrix)
.br
	XpexMatrix3X3 transformation_matrix;  
.br
	XpexCoord2D *fixed_point;
.br
	XpexVector2D *shift_vector;  
.br
	XpexFloat rotation_angle;  
.br
	XpexVector2D *scale_vector;  
.br
	XpexMatrix3X3 composed_matrix; /* RETURN */ 
.sp
int XpexComputeGeometricNormal(vert_attr, color_type, vertex_lists, num_lists, 
.br
	geometric_normal)
.br
	XpexBitmaskShort vert_attr;
.br
	XpexColorType color_type;
.br
	XpexListOfVertex *vertex_lists;
.br
	int num_lists;
.br
	XpexVector3D *geometric_normal; /* RETURN */
.fi
.SH "Arguments"
.IP \fIvert_attr\fP 1i
Specifies what vertex data is available for each vertex 
of the primitive.
.IP \fIvertex_lists\fP 1i
Specifies the fill area data.
.IP \fInum_lists\fP 1i
Specifies the number of fill areas in the set.
.IP \fIgeometric_normal\fP 1i
Specifies a pointer to the location to store the computed 
geometric normal.
.SH "Description"
This function computes the geometric normal of the fill area set.
.SH Structures
To be added 
.SH "Errors"
If the return value is 1, the fill area is degenerate.
Otherwise, the default return value is 0.
.SH "Related Commands"
XpexFillAreaSet, XpexExtFillAreaSet,
XpexSetLocalTransform, XpexSetGlobalTransform,
XpexSetLocalTransform2D, XpexSetGlobalTransform2D.


contrib/lib/Xpex/extensions/lib/PEX/Xpex/lib/utilities/XpexCompose.c
/* $Header: XpexCompose.c,v 2.2 91/09/11 15:51:16 sinyaw Exp $ */
/* 
 *	XpexCompose.c
 */

/*****************************************************************
Copyright 1991 by Sony Microsystems Company, San Jose, California

					All Rights Reserved

Permission to use, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted,
provided that the above copyright notice appear in all copies and
that both that copyright notice and this permission notice appear
in supporting documentation, and that the name of Sony not be used
in advertising or publicity pertaining to distribution of the
software without specific, written prior permission.

SONY DISCLAIMS ANY AND ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
INCLUDING ALL EXPRESS WARRANTIES AND ALL IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS, FOR A PARTICULAR PURPOSE. IN NO EVENT
SHALL SONY BE LIABLE FOR ANY DAMAGES OF ANY KIND, INCLUDING BUT NOT
LIMITED TO SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES RESULTING FROM
LOSS OF USE, DATA OR LOSS OF ANY PAST, PRESENT, OR PROSPECTIVE PROFITS,
WHETHER IN AN ACTION OF CONTRACT, NEGLIENCE OR OTHER TORTIOUS ACTION, 
ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS 
SOFTWARE.
*****************************************************************/

#include <stdio.h>
#include <math.h>
#include "Xpexlib.h"

/*
 * Multiply matrices: m = a * b;
 *
 * This simple-minded explicit form is about 40% faster than
 * nested loops and indexes.
 */

static void
_XpexMatMul(m , a, b)
	XpexMatrix	m;
	XpexMatrix	a;
	register	XpexMatrix	b;
{
    register XpexFloat	*r;
    register int	i;

    
    if ((m != a) && (m != b)) {
	for (i = 0; i < 4; i++) {
	    r = a[i];
	    m[i][0] = r[0]*b[0][0] + r[1]*b[1][0] + r[2]*b[2][0] +r[3]*b[3][0];
	    m[i][1] = r[0]*b[0][1] + r[1]*b[1][1] + r[2]*b[2][1] +r[3]*b[3][1];
	    m[i][2] = r[0]*b[0][2] + r[1]*b[1][2] + r[2]*b[2][2] +r[3]*b[3][2];
	    m[i][3] = r[0]*b[0][3] + r[1]*b[1][3] + r[2]*b[2][3] +r[3]*b[3][3];
	}
    }
    else {
		XpexMatrix	t;
		register	XpexFloat       *tp, *mp;
	
	for (i = 0; i < 4; i++) {
	    r = a[i];
	    t[i][0] = r[0]*b[0][0] + r[1]*b[1][0] + r[2]*b[2][0] +r[3]*b[3][0];
	    t[i][1] = r[0]*b[0][1] + r[1]*b[1][1] + r[2]*b[2][1] +r[3]*b[3][1];
	    t[i][2] = r[0]*b[0][2] + r[1]*b[1][2] + r[2]*b[2][2] +r[3]*b[3][2];
	    t[i][3] = r[0]*b[0][3] + r[1]*b[1][3] + r[2]*b[2][3] +r[3]*b[3][3];
	}
	tp = &t[0][0];
	mp = &m[0][0];
	for (i = 0; i < 16; i++)
	    *mp++ = *tp++;
    }
}

static void
_XpexMatMul3X3(m, a, b)
	XpexMatrix3X3	m;	/* OUT result matrix	*/
	XpexMatrix3X3	a;      /* matrix a	*/
	register	XpexMatrix3X3	b;	/* matrix b	*/
{
    register float	*r;
    register int	i;

    if ((m != a) && (m != b)) {
	for (i = 0; i < 3; i++) {
	    r = a[i];
	    m[i][0] = r[0]*b[0][0] + r[1]*b[1][0] + r[2]*b[2][0];
	    m[i][1] = r[0]*b[0][1] + r[1]*b[1][1] + r[2]*b[2][1];
	    m[i][2] = r[0]*b[0][2] + r[1]*b[1][2] + r[2]*b[2][2];
	}
    }
    else {
		XpexMatrix3X3	t;
		register	float	*tp, *mp;
	
	for (i = 0; i < 3; i++) {
	    r = a[i];
	    t[i][0] = r[0]*b[0][0] + r[1]*b[1][0] + r[2]*b[2][0];
	    t[i][1] = r[0]*b[0][1] + r[1]*b[1][1] + r[2]*b[2][1];
	    t[i][2] = r[0]*b[0][2] + r[1]*b[1][2] + r[2]*b[2][2];
	}
	tp = &t[0][0];
	mp = &m[0][0];
	for (i = 0; i < 9; i++)
	    *mp++ = *tp++;
    }
}

void
XpexComposeMatrix(a, b, m)
    XpexMatrix    a;              /* matrix a     */
    XpexMatrix    b;              /* matrix b     */
    XpexMatrix    m;              /* OUT result matrix    */
{
    _XpexMatMul(m, a, b);
}

void
XpexComposeMatrix2D(a, b, m)
    XpexMatrix3X3     a;              /* matrix a     */
    XpexMatrix3X3     b;              /* matrix b     */
    XpexMatrix3X3     m;              /* OUT result matrix    */
{
    _XpexMatMul3X3(m, a, b);
}

void
XpexComposeTransformationMatrix(m, 
	pt, shift, x_ang, y_ang, z_ang, scale, 
	result)
    XpexMatrix      m;              /* transformation matrix        */
    XpexCoord3D     *pt;            /* fixed point  */
    XpexVector3D    *shift;         /* shift vector */
    XpexFloat       x_ang;          /* rotation angle X     */
    XpexFloat       y_ang;          /* rotation angle Y     */
    XpexFloat       z_ang;          /* rotation angle Z     */
    XpexVector3D    *scale;         /* scale vector */
    XpexMatrix      result;         /* OUT transformation matrix    */
{
	extern void _XpexBuildTransform();

    XpexMatrix    xform;

    _XpexBuildTransform(pt, shift, x_ang, y_ang, z_ang, scale, xform);
    /* Assuming pre-multiplication of old by new. */
    _XpexMatMul(result, xform, m);
}

void
XpexComposeTransformationMatrix2D(m, pt, shift, angle, scale, 
	result)
    XpexMatrix3X3    m;      /* transformation matrix        */
    XpexCoord2D      *pt;    /* fixed point  */
    XpexVector2D     *shift; /* shift vector */
    XpexFloat        angle;  /* rotation angle       */
    XpexVector2D     *scale; /* scale vector */
    XpexMatrix3X3    result; /* OUT transformation matrix    */
{
	extern void _XpexBuildTransform2D();

    XpexMatrix3X3     xform;

    _XpexBuildTransform2D(pt, shift, angle, scale, xform);
    /* Assuming pre-multiplication of old by new. */
    _XpexMatMul3X3(result, xform, m);
}
